src/testFixtures/java/org/opensearch/neuralsearch/BaseNeuralSearchIT.java

```
@@ -1504,23 +1505,32 @@ protected void wipeOfTestResources(
         final String modelId,
         final String searchPipeline
     ) throws IOException {
-        if (ingestPipeline != null) {
-            deletePipeline(ingestPipeline);
-        }
-        if (searchPipeline != null) {
-            deleteSearchPipeline(searchPipeline);
-        }
-        if (modelId != null) {
-            try {
-                deleteModel(modelId);
-            } catch (AssertionError e) {
-                // sometimes we have flaky test that the model state doesn't change after call undeploy api
-                // for this case we can call undeploy api one more time
-                deleteModel(modelId);
+        try {
+            if (ingestPipeline != null) {
+                deletePipeline(ingestPipeline);
+            }
+            if (searchPipeline != null) {
+                deleteSearchPipeline(searchPipeline);
+            }
+            if (modelId != null) {
+                try {
+                    deleteModel(modelId);
+                } catch (AssertionError e) {
+                    // sometimes we have flaky test that the model state doesn't change after call undeploy api
+                    // for this case we can call undeploy api one more time
+                    deleteModel(modelId);
+                }
+            }
+            if (indexName != null) {
+                deleteIndex(indexName);
+            }
+        } catch (ResponseException e) {
+            // It's possible that test fails during resources (index, model, pipeline, etc.) creation, when clean up
+            // these resources after test run, the delete methods will throw ResponseException with 404 Not Found code
+            // In this case, we just need to ignore this exception, for other exceptions, continue throwing
+            if (RestStatus.fromCode(e.getResponse().getStatusLine().getStatusCode()) != RestStatus.NOT_FOUND) {
+                throw e;
```

This exception will still obscure the original exception. Therefore, I believe resource cleanup should be managed in the `@After` method rather than in the `finally` block.

---

So are you saying we should shift this cleanUp at every integ class level?

---

When I developed the framework you suggested to do it in in current way https://github.com/opensearch-project/neural-search/pull/533#discussion_r1465646182

---

cleanUp at every individual test level but inside base class so that each test class don't repeat the same code. https://github.com/opensearch-project/k-NN/blob/main/src/testFixtures/java/org/opensearch/knn/ODFERestTestCase.java#L152

---

So it will mean just removal of code under finally block and move it in @After. The cleanUp method will go to base class so it in a particular integ class it will look little better. But underlying work still remains the same

---

Is my understand correct?

---

Yes. However, there is one more difference. When it fails inside @after method, it won't obscure the exception thrown during test I believe?

---

> This exception will still obscure the original exception.

Yes, the original exception was overriden by the exception that thrown in the finally block, hence we were not able to see the original exception in the first place. 

I will make more code change to remove the finally block and add it to an @After method

---

Even if we move to @after block the root issue is that we shouldn't be calling this method passing in resources we haven't created yet right? The method is still called incorrectly, we just hide the exception?

---

> Even if we move to @after block the root issue is that we shouldn't be calling this method passing in resources we haven't created yet right? The method is still called incorrectly, we just hide the exception?

You are right. When we moved to @after block, we won't depend on static value of resource name but we will make an api call to list all available resources and clean them up.